### PR TITLE
Fix/get top fashion cities names

### PR DESCRIPTION
**Pull request by devsadeq**

### CODE REVIEW

**Review by flutterdevtaha — APPROVED** [review]
well done 🥰

**Review by MahmoudAkramAlMallahi — APPROVED** [review]

File: src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt [hunk 1]
```diff
@@ -9,21 +10,25 @@ import org.junit.jupiter.api.function.Executable
...
     fun setUp() {
         fakeDataSource = FakeDataSource()
-        getTopFashionCitiesNamesInteractor = GetTopFashionCitiesNamesInteractor(fakeDataSource)
+        emptyFakeDataSource = FakeDataWithEmptyList()
+        interactor = GetTopFashionCitiesNamesInteractor(fakeDataSource)
+        emptyInteractor = GetTopFashionCitiesNamesInteractor(emptyFakeDataSource)
     }
 
     @Test
     fun `should return only the required number of results`() {
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt:27]:
```suggestion
    fun `should return only the required number of cities`() {
```

File: src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt [hunk 2]
```diff
@@ -9,21 +10,25 @@ import org.junit.jupiter.api.function.Executable
...
         // given limit
         val limit = 5
         // when get top 5 cities names
-        val actual = getTopFashionCitiesNamesInteractor.execute(limit).size
+        val actual = interactor.execute(limit).size
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt:31]: Are you sure about this test case?
Is it make sense?

File: src/main/kotlin/utils/Constants.kt [hunk 3]
```diff
@@ -2,6 +2,5 @@ package utils
 
 object Constants {
```
Comment (ma28alhams) [inline src/main/kotlin/utils/Constants.kt:3]: use **companion object** instead of object

**Review by ma28alhams — CHANGES_REQUESTED** [review]
- Add a test case to check if all the data has been filtered, then should return an empty list.

File: src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt [hunk 4]
```diff
@@ -13,10 +14,11 @@ class GetTopFashionCitiesNamesInteractor(
     fun execute(limit: Int): List<String> {
         return dataSource
             .getAllCitiesData()
+            .ifEmpty { throw NoReturnedDataException(EMPTY_DATA_SOURCE_EXCEPTION_MSG) }
```
[the remark this answers is not in the document]
Comment (devsadeq) [reply src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt]: Already done

File: src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt [hunk 5]
```diff
@@ -13,10 +14,11 @@ class GetTopFashionCitiesNamesInteractor(
...
             .asSequence()
             .filter(::excludeNullPricesAndLowQualityData)
             .sortedBy { it.clothesPrices.average() }
-            .distinctBy { Pair(it.cityName, it.country) }
+            .distinctBy { it.cityName to it.country }
```
[the remark this answers is not in the document]
Comment (devsadeq) [reply src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt:21]: To get a unique cities, no duplications

File: src/main/kotlin/interactor/GetCityMatchManagerExpectationsInteractor.kt [hunk 6]
```diff
@@ -10,7 +10,7 @@ class GetCityMatchManagerExpectationsInteractor(
 
     fun execute(): CityEntity? {
         return dataSource.getAllCitiesData()
-            .let { it.ifEmpty { throw NoReturnedDataException(Constants.EMPTY_LIST_EXCEPTION) } }
+            .let { it.ifEmpty { throw NoReturnedDataException(EMPTY_LIST_EXCEPTION_MSG) } }
             .filter { it.country in onlyRequiredCountries }
             .let { it.ifEmpty { emptyList() } }
             .minByOrNull(::getMidRangePrices)
```
[the remark this answers is not in the document]
Comment (devsadeq) [reply src/main/kotlin/interactor/GetCityMatchManagerExpectationsInteractor.kt:16]: This code is not mine. I just resolve a conflict here. It's belong to Taha.

[the remark this answers is not in the document]
Comment (devsadeq) [reply src/main/kotlin/interactor/GetCityMatchManagerExpectationsInteractor.kt:16]: This also belong to @flutterdevtaha 😅

File: src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt (hunk 2 above)
Comment (devsadeq) [reply src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt:31]: Actually, this text case save me several times so I think I need it.

File: src/main/kotlin/utils/Constants.kt (hunk 3 above)
Comment (devsadeq) [reply src/main/kotlin/utils/Constants.kt:3]: How to do that exactly?

https://github.com/Salmon-family/Clubs/blob/master/core/ui/src/main/java/com/devfalah/ui/util/Constants.kt

**Comment by devsadeq** [comment]
I'm not get it 👀 @ma28alhams

File: src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ma28alhams) [reply src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt:21]: Why distincting by **Pair** ?

File: src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt (hunk 2 above)
Comment (ma28alhams) [reply src/test/kotlin/interactor/GetTopFashionCitiesNamesInteractorTest.kt:31]: I think Bariq said that we should not compare the size of the arrays to each other, but rather compare the two arrays.

File: src/main/kotlin/utils/Constants.kt (hunk 3 above)
Comment (ma28alhams) [reply src/main/kotlin/utils/Constants.kt:3]: Object used as Singelton pattern,
But companion object used to declare members as static members.

https://medium.com/mindorks/kotlin-object-vs-companion-object-a1907c76a2af

File: src/main/kotlin/interactor/GetCityMatchManagerExpectationsInteractor.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (flutterdevtaha) [reply src/main/kotlin/interactor/GetCityMatchManagerExpectationsInteractor.kt:16]: I will edit on @ma28alhams

File: src/main/kotlin/utils/Constants.kt (hunk 3 above)
Comment (flutterdevtaha) [reply src/main/kotlin/utils/Constants.kt:3]: @ma28alhams that is more correct becuase it is shared across the app

File: src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (devsadeq) [reply src/main/kotlin/interactor/GetTopFashionCitiesNamesInteractor.kt:21]: Cause there are many counties that have same name of cities.
